### PR TITLE
 xreplace

### PR DESCRIPTION
**Pull request by ranjithkumar007**
cc @isuruf.

### CODE REVIEW

File: symengine/subs.h [hunk 1]
```diff
@@ -206,13 +209,121 @@ class SubsVisitor : public BaseVisitor<SubsVisitor>
...
+{
+public:
+    using XReplaceVisitor::bvisit;
+
+    SubsVisitor(const map_basic_basic &xreplace_dict_)
+        : BaseVisitor<SubsVisitor, XReplaceVisitor>(xreplace_dict_)
+    {
+    }
+
     void bvisit(const Derivative &x)
```
Comment (isuruf) [inline symengine/subs.h]: You should move this to `xreplace`. See sympy's implementation

File: symengine/subs.h [hunk 2]
```diff
@@ -252,8 +363,8 @@ class SubsVisitor : public BaseVisitor<SubsVisitor>
                         break;
                     }
                 } else {
-                    result_
-                        = make_rcp<const Subs>(x.rcp_from_this(), subs_dict_);
+                    result_ = make_rcp<const Subs>(x.rcp_from_this(),
```
Comment (ranjithkumar007) [inline symengine/subs.h]: @isuruf, If `Derivative` is moved to `xreplace`,  does returning an instance of `Subs` class on a call to `xreplace` make sense?

Comment (isuruf) [inline symengine/subs.h]: Yes. SymPy does it as well.

Comment (ranjithkumar007) [inline symengine/subs.h]: Ah. Fine then

**Comment by ranjithkumar007** [comment]
ping @isuruf

File: symengine/subs.h [hunk 3]
```diff
@@ -11,14 +11,16 @@ RCP<const Basic> msubs(const RCP<const Basic> &x,
...
                        const map_basic_basic &subs_dict);
 
-class SubsVisitor : public BaseVisitor<SubsVisitor>
+class XReplaceVisitor : public BaseVisitor<XReplaceVisitor>
 {
+
 protected:
     RCP<const Basic> result_;
-    const map_basic_basic &subs_dict_;
+    const map_basic_basic &xreplace_dict_;
```
Comment (isuruf) [inline symengine/subs.h]: Let's leave `subs_dict_` as it was.

File: symengine/subs.h [hunk 4]
```diff
@@ -407,6 +410,50 @@ inline RCP<const Basic> ssubs(const RCP<const Basic> &x,
...
+
+    void bvisit(const Pow &x)
+    {
+        RCP<const Basic> base_new = apply(x.get_base());
+        RCP<const Basic> exp_new = apply(x.get_exp());
+        if (xreplace_dict_.size() == 1
+            and is_a<Pow>(*((*xreplace_dict_.begin()).first))
+            and not is_a<Add>(
+                    *down_cast<const Pow &>(*(*xreplace_dict_.begin()).first)
+                         .get_exp())) {
```
Comment (isuruf) [inline symengine/subs.h]: What happens to `(2**(x+y+z)).subs({2**(x+z): 1})`?

File: symengine/subs.h [hunk 5]
```diff
@@ -407,6 +410,50 @@ inline RCP<const Basic> ssubs(const RCP<const Basic> &x,
...
+{
+public:
+    using XReplaceVisitor::bvisit;
+
+    SubsVisitor(const map_basic_basic &xreplace_dict_)
+        : BaseVisitor<SubsVisitor, XReplaceVisitor>(xreplace_dict_)
+    {
+    }
+
+    void bvisit(const Pow &x)
```
Comment (isuruf) [inline symengine/subs.h]: Only `Pow`? Isn't `Add` and `Mul` also needed?

File: symengine/subs.h (hunk 4 above)
Comment (ranjithkumar007) [inline symengine/subs.h]: `2**(x+y+z)`

File: symengine/subs.h (hunk 5 above)
Comment (ranjithkumar007) [inline symengine/subs.h]: Needed.

Comment (ranjithkumar007) [inline symengine/subs.h]: on second thought, `Add` and `Mul` are not performing any such simplifications. So, It might be better to keep it as it is.

**Comment by ranjithkumar007** [comment]
ping @isuruf

File: symengine/subs.h (hunk 4 above)
Comment (ranjithkumar007) [inline symengine/subs.h]: even in SymPy.

**Comment by ranjithkumar007** [comment]
can you please have a look @Sumith1896 ?

**Pull request merged**